design-documents/csp.md

```diff
@@ -0,0 +1,169 @@
+along with disabling certain unsafe HTML features like inline JavaScript. It's a powerful tool
+in the fight with XSS attacks.
+## Modes
+CSP supports two modes:
+* restrict mode `Content-Security-Policy`
+* report violations mode `Content-Security-Policy-Report-Only`
+ 
+1st one would follow defined rules and prevent certain scenarios from happening
+(like not loading a font from an untrusted address).
+ 
+Second one will only report violations
+to the server but proceed as it would without these restrictions. This mode would be convenient to merchants
+just adopting CSP on their existing stores to see what kind of legit resources/features they need to enable
+via CSP.
+## Backward compatibility
+In order to remind merchants that having CSP defined is important it will be enabled by default. But since adopting it
+would require manually whitelisting trusted resource/features by merchants and extension developers CSP
+will be functioning in the report-only mode. This will allow us to introduce CSP in a patch release.
+ 
+## API
+At the beginning Magento would not manage all of the possible CSP policies, that's
+why we need an extensible SPI to allow 3rd party developers to implement and control new
+policies.
+ 
+CSP API/SPI will consist of the following interfaces:
+ 
+* Policy DTO. The interface has only policy ID because not all CSP policies look like _\<policy name\> \<value\> \<value\>_,
+  some may consist from multiple headers and have different syntax
+```php
+namespace Magento\Security\Model\CSP\Data;
+
+interface PolicyInterface
+{
+    //Policy ID, like 'default-src' or 'referrer'
+    public function getId(): string;
+}
+```
+ 
+* Policy collector, will collect policies from different sources, will be composed from multiple collectors.
+```php
+namespace Magento\Security\Model\CSP;
+
+interface PolicyCollectorInterface
+{
+    //Keys - policy IDs.
+    public function collect(PolicyInterface[] $defaultPolicies = []): PolicyInterface[];
+}
+```
+ 
+* Headers generator, will create headers based on a policy provided
+```php
+namespace Magento\Security\Model\CSP;
+
+interface PolicyHeadersGeneratorInterface
+{
+    //With keys as header names and values as header values.
+    public function generateHeaders(PolicyInterface $policy): array
+}
+```
+ 
+* A violation recorded by a browser
+```php
+namespace Magento\Security\Model\CSP\Data;
+
+interface ViolationInterface
+{
+    public function getDocumentUri(): string;
+    
+    public function getViolatedDirective(): string;
+    
+    public function getOriginalPolicy(): string;
+    
+    public function getReferrer(): ?string;
+    
+    public function getBlockedUri(): ?string;
+    
+    public function getSourceJsFileLine(): ?string;
+    
+    public function getReported(): \DateTime
+}
+```
+ 
+* A repository of violations logs collected when CSP reporting is enabled
```

Comment:
Need to think how to make reporting works. I don't think we can't use Web API to capture the data, it would be just too much load. May need to build a separate service for capturing reports data.

Reply:
we can use a regular controller as the report endpoint for browsers if our web API mechanism would be difficult to fit for this